source/differentiation/ad/ad_helpers.cc

```diff
@@ -1125,6 +1125,173 @@ namespace Differentiation
+      symmetric_independent_variables =
+        std::vector<bool>(new_n_independent_variables, false);
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    bool
+    ADHelperPointLevelFunctionsBase<dim, ADNumberTypeCode, ScalarType>::
+      is_symmetric_independent_variable(const unsigned int index) const
+    {
+      Assert(index < symmetric_independent_variables.size(),
+             ExcInternalError());
+      return symmetric_independent_variables[index];
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    unsigned int
+    ADHelperPointLevelFunctionsBase<dim, ADNumberTypeCode, ScalarType>::
+      n_symmetric_independent_variables() const
+    {
+      return std::count(symmetric_independent_variables.begin(),
+                        symmetric_independent_variables.end(),
+                        true);
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperPointLevelFunctionsBase<dim, ADNumberTypeCode, ScalarType>::
+      register_independent_variables(const std::vector<scalar_type> &values)
+    {
+      // This is actually the same thing the set_independent_variable function,
+      // in the sense that we simply populate our array of independent values
+      // with a meaningful number. However, in this case we need to double check
+      // that we're not registering these variables twice
+      Assert(values.size() == this->n_independent_variables(),
+             ExcMessage(
+               "Vector size does not match number of independent variables"));
+      for (unsigned int i = 0; i < this->n_independent_variables(); ++i)
+        {
+          Assert(this->registered_independent_variable_values[i] == false,
+                 ExcMessage("Independent variable value already registered."));
+        }
+      set_independent_variables(values);
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    const std::vector<
+      typename ADHelperPointLevelFunctionsBase<dim,
+                                               ADNumberTypeCode,
+                                               ScalarType>::ad_type> &
+    ADHelperPointLevelFunctionsBase<dim, ADNumberTypeCode, ScalarType>::
+      get_sensitive_variables()
+    {
+      if (ADNumberTraits<ad_type>::is_taped == true)
+        {
+          Assert(this->active_tape_index() !=
+                   Numbers<ad_type>::invalid_tape_index,
+                 ExcMessage("Invalid tape index"));
+        }
+
+      // If necessary, initialize the internally stored vector of
+      // AD numbers that represents the independent variables
+      this->finalize_sensitive_independent_variables();
```

Comment:
what does the "if necessary" refer to? The call in line 1241 is unconditional, so I figure that it's a dependence in the base class whether the function does or does not do anything?